### PR TITLE
fix(upgrade): populate upgraded component's view before creating the controller

### LINKED ISSUE #13912: Upgrading a Ng1Component and using $element does not work
**I'm submitting a ...**  (check one with "x")
```
[x] bug report 
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Trying to upgrade a ng1 component that uses inline templates and the $element service. The angular2 component will have NO children nodes.

**Expected behavior**
The child nodes should be available instantly (like in ng1)

**Minimal reproduction of the problem with instructions**
I'm running a ng1 app and use ngUpgrade to upgrade ng1 components to ng2 ones. The inline template is not instantly available.

*MINIMAL DEMO*
```
(function () {
    'use strict';

    angular.module('testComponent', []);

    angular.module('testComponent')
        .component('testComponent', {
            template: '<div><p>This is shitty</p></div>',
            controller: testComponentController
        })

    function testComponentController($element, $timeout) {
        console.log('Instant', $element.children());
        this.$onInit = function () {
            console.log('On Init', $element.children());
            $timeout(function () {
                console.log('After Timeout', $element.children());
            });
        }
    }
}
```

*OUTPUT NG1*
```
Instant [div]
On Init [div]
After Timeout [div]
```

*OUTPUT NG2*
```
Instant JQLite {}
On Init JQLite {}
After Timeout [div.ng-scope]
```

As you can see, the element is not available until the $timeout.

**What is the motivation / use case for changing the behavior?**
Gradually upgrade a ng1 application to ng2 instead of having to do a big bang

**Please tell us about your environment:**
Cordova android/ios/windows Phone hybrid angular 1 app trying to migrate to angular2
Using: 
- typescript 2.1.4
- Angular 2.4.1
- Angular 1.5.8
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
Typescript compiling to ES5

### PR DESCRIPTION
**Pull request by gkalpak**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
The relative order of the AngularJS compiling/linking operations is not similar to AngularJS's, resulting in inconsistent behavior for upgraded components (which makes upgrading to Angular less straight forward).

**What is the new behavior?**
The compiling/linking process of AngularJS is followed more closely.
Main differences:

- The components view is already populated when the controller is instantiated (and subsequent hooks are called).
- The correct DOM content is available when running the `$onChanges`, `$onInit`, `$doCheck` hooks.
   Previously, the "content children" were still present, not the "view children".
- The same for pre-linking.
- The template is compiled in the correct DOM context (e.g. has access to ancestors).
   Previously, it was compiled in isolation, inside a dummy element.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

**Other information**:
Fixes #13912.

For reference, here is the order of operations:

**Before**

1. Compile template
2. Instantiate controller
3. Hook: $onChanges
4. Hook: $onInit
5. Hook: $doCheck
6. Pre-linking
7. Collect content children
8. Insert compiled template
9. Linking
10. Post-linking
11. Hook: $postLink

**After**

1. Collect content children
2. Insert template
3. Compile template
4. Instantiate controller
5. Hook: $onChanges
6. Hook: $onInit
7. Hook: $doCheck
8. Pre-linking
9. Linking
10. Post-linking
11. Hook: $postLink

### CODE REVIEW

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts [hunk 1]
```diff
@@ -1020,6 +1020,54 @@ export function main() {
          }));
     });
 
+    describe('compiling', () => {
+      it('should compile the ng1 template in the correct DOM context', async(() => {
+           let grantParentNodeName: string;
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: should be `grandParentNodeName`?

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts [hunk 2]
```diff
@@ -1020,6 +1020,54 @@ export function main() {
...
+                                 .component('ng1A', ng1ComponentA)
+                                 .directive('ng1B', () => ng1DirectiveB)
+                                 .directive('ng2X', downgradeComponent({component: Ng2ComponentX}));
+
+           // Define `Ng2Module`
+           @NgModule({
+             imports: [BrowserModule, UpgradeModule],
+             declarations: [Ng1ComponentAFacade, Ng2ComponentX],
+             entryComponents: [Ng2ComponentX],
+             schemas: [NO_ERRORS_SCHEMA],
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: We don't need `NO_ERRORS_SCHEMA`

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts [hunk 3]
```diff
@@ -1253,6 +1301,55 @@ export function main() {
              expect(multiTrim(element.textContent)).toBe('WORKS GREAT');
            });
          }));
+
+      it('should insert the compiled content before instantiating the controller', async(() => {
+           let compiledContent: string;
+
+           // Define `ng1Component`
+           const ng1Component: angular.IComponent = {
+             template: 'Hello, world!',
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: Given that you refer to "compiled" content here, is it worth making the template contain something dynamic, like an interpolation, rather than just a static string?

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts [hunk 4]
```diff
@@ -1784,13 +1881,8 @@ export function main() {
...
              controllerAs: '$ctrl',
-             controller: class {
-               constructor($scope: angular.IScope) {
-                 Object.getPrototypeOf($scope)['$onChanges'] = scopeOnChanges;
-               }
-
-               $onChanges(changes: SimpleChanges) { controllerOnChangesB(changes); }
-             }
+             controller:
+                 class {$onChanges(changes: SimpleChanges) { controllerOnChangesB(changes); }}
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: Is this what our code formatter did to this line? Ugly!

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts [hunk 5]
```diff
@@ -1827,7 +1919,10 @@ export function main() {
            const ng1Module = angular.module('ng1Module', [])
                                  .directive('ng1A', () => ng1DirectiveA)
                                  .directive('ng1B', () => ng1DirectiveB)
-                                 .directive('ng2', downgradeComponent({component: Ng2Component}));
+                                 .directive('ng2', downgradeComponent({component: Ng2Component}))
+                                 .run(($rootScope: angular.IRootScopeService) => {
+                                   Object.getPrototypeOf($rootScope)['$onChanges'] = scopeOnChanges;
+                                 });
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: What's going on here. Why this change?

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 6]
```diff
@@ -413,19 +435,28 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
 
       // QUESTION: Do we want the ng1 component to call the function with `<value>` or with
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: I think we can remove this question now. It looks like we have agreed upon the former.

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 7]
```diff
@@ -321,6 +331,18 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
     return bindings;
   }
 
+  private getContentChildNodes(element: Element): Node[] {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: Since this function has a side effect (of removing nodes from the `element`) then `get...` is not a good name, since it makes it sound like a pure getter function. Perhaps `extractContentChildNodes` or just `extractChildNodes` would be more appropriate?

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Not quite finished but here is my review so far.

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 8]
```diff
@@ -137,24 +137,33 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
     // QUESTION 2: Should we make the scope accessible through `$element.scope()/isolateScope()`?
     this.$componentScope = $parentScope.$new(!!this.directive.scope);
 
+    this.tempBindingDestination = Object.create(null);
+    this.setupOutputs(this.tempBindingDestination);
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: Given that `setupOutputs` is only called here, why not change it to simply return this object and call it something like `createOutputs(): IBindingDestination`. So these two lines would become something like:

```ts
this.tempOutputHolder = this.setupOutputs();
```

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 9]
```diff
@@ -137,24 +137,33 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
...
       throw new Error(
-          `Upgraded directive '${name}' specifies 'bindToController' but no controller.`);
+          `Upgraded directive '${this.directive.name}' specifies 'bindToController' but no controller.`);
     }
 
+    // Set up outputs
     this.bindingDestination = bindToController ? this.controllerInstance : this.$componentScope;
+    Object.keys(this.tempBindingDestination)
+        .forEach(key => this.bindingDestination[key] = this.tempBindingDestination[key]);
+    this.tempBindingDestination = null;
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: Then this bit of code would become something like:

```ts
    Object.keys(this.tempOutputHolder)
        .forEach(key => this.bindingDestination[key] = this.tempOutputHolder[key]);
    this.tempOutputHolder = null;
```

(It is a shame we probably cannot use `Object.assign` here.)

File: modules/@angular/upgrade/src/static/upgrade_component.ts (hunk 8 above)
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: This would IMO make it easier to see what is going on here - i.e. that we need to be able to handle changes before the outputs get assigned to the binding destination (right?).

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 10]
```diff
@@ -95,10 +95,11 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
...
   private directive: angular.IDirective;
   private bindings: Bindings;
-  private linkFn: angular.ILinkFn;
 
-  private controllerInstance: IControllerInstance = null;
-  private bindingDestination: IBindingDestination = null;
+  private controllerInstance: IControllerInstance;
+  private bindingDestination: IBindingDestination;
+  private tempBindingDestination: IBindingDestination;
+  private pendingChanges: SimpleChanges;
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: I think some comments about these two new properties to explain their purpose to future readers would be helpful here.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
The behaviour seems correct. I have a few minor suggestions about the implementation, naming and commenting.

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts (hunk 4 above)
Comment (gkalpak) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: Apparently, it is both a formater and an obfuscator :persevere:

File: modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts (hunk 5 above)
Comment (gkalpak) [inline modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts]: We need this to happen before `ngOnChanges` is run. Previously (as a convenience) I was doing this inside the controller, but now that the controller is instantiated after `ngOnChanges`, it is to late to patch `Scope.prototype`. So I moved it into a `run` block.

**Comment by gkalpak** [comment]
@petebacondarwin: PTAL
I addressed your comments and made another change redarding `setupOutputs`.
I realized we were doing two thing in `setupOutpus`:

1. Define `EventEmitters` for the outputs on the ng2 facade compoenent. (This needed to happen in the component constructor, or Angular would complain).
2. Set up the forwarding of the `&` bindings from the ng1 component to the ng2 component facade's `EventEmitter`. (This is not needed in the constructor, but it requires the binding destination.)

So, I kept the first part in the constructor and moved the second part to `ngOnInit` (after the ng1 controller has been created).

**Comment by gkalpak** [comment]
(The Travis failure is unrelated btw. Inherited it from master.)

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 11]
```diff
@@ -95,10 +95,13 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
...
   private bindings: Bindings;
-  private linkFn: angular.ILinkFn;
 
-  private controllerInstance: IControllerInstance = null;
-  private bindingDestination: IBindingDestination = null;
+  private controllerInstance: IControllerInstance;
+  private bindingDestination: IBindingDestination;
+
+  // We will be instantiating the controller in the `ngOnInit` hook, when the first `ngOnChanges`
+  // will have been already triggered. We store the `SimpleChanges` and to "play them back" later.
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: > We store the `SimpleChanges` and ~~to~~ "play them back" later.

Remove the "to"

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 12]
```diff
@@ -405,32 +426,42 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
   }
 
   private setupOutputs() {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: could we call this `createOutputs` or something similar?

File: modules/@angular/upgrade/src/static/upgrade_component.ts [hunk 13]
```diff
@@ -405,32 +426,42 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
...
-    });
+    // Set up the outputs for `=` and `&` bindings
+    this.bindings.twoWayBoundProperties.concat(this.bindings.expressionBoundProperties)
+        .forEach(propName => {
+          const outputName = this.bindings.propertyToOutputMap[propName];
+          (this as any)[outputName] = new EventEmitter();
+        });
+  }
 
+  private setupExpressionBindings() {
```
Comment (petebacondarwin) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: And then call this `bindOutputs` or something similar?

**Review by petebacondarwin — APPROVED** [review]
I made some naming suggestions but they are not blockers on landing this.

File: modules/@angular/upgrade/src/static/upgrade_component.ts (hunk 12 above)
Comment (gkalpak) [inline modules/@angular/upgrade/src/static/upgrade_component.ts]: How about `initializeOutputs()` (as it basically initializes the output properties to `new EventEmitter()`)?

**Comment by gkalpak** [comment]
I addressed the comments and squashed.

**Comment by petebacondarwin** [comment]
LGTM
